### ISSUE TITLE
Improve lore2sphinx buildbot results for `projects/core/development`

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4567]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)| @khorn reported|
|-|-|
|Trac ID|trac#4567|
|Type|enhancement|
|Created|2010-07-11 06:43:37Z|
|Branch|https://github.com/twisted/twisted/tree/improve-lore2sphinx-output-for-projects-core-development-4567|

Make the lore2sphinx buildbot results for `projects/core/development` look as good as possible by modifying the lore sources.

Things to look for:
 1. nested inline markup should be removed
 1. whitespace should be added or removed as needed to make ReST markup render properly
 1. anything else that you can improve by modifying the Lore sources

Note that you probably want to run lore2sphinx on a local copy for development purposes.

You can get lore2sphinx from:
http://bitbucket.org/khorn/lore2sphinx

The lore2sphinx buildbot results can be found at:

http://buildbot.twistedmatrix.com/builds/sphinx-html-15615/

(or at different directories for different revisions...the containing directory is browseable)

Attachments:

* [lore-core-development-cleanups.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/837/83787f060a59493aefdcd4b2369990e7303e186e/c49be266d7c0e065701597c95359b7854193aa32.diff) (18990 bytes) - added by Screwtape on 2010-07-11 10:51:08Z - Changes that make the lore2sphinx conversion cleaner.

<details><summary>Searchable metadata</summary>

```
trac-id__4567 4567
type__enhancement enhancement
reporter__khorn khorn
priority__normal normal
milestone__Lore_to_Sphinx Lore to Sphinx
branch__branches_improve_lore2sphinx_output_for_projects_core_development_4567 branches/improve-lore2sphinx-output-for-projects-core-development-4567
branch_author__screwtape screwtape
status__closed closed
resolution__fixed fixed
component__core core
keywords__documentation documentation
time__1278830617000000 1278830617000000
changetime__1300088151000000 1300088151000000
version__None None
owner__lvh lvh
cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="Screwtape's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Screwtape set owner to Screwtape|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="Screwtape's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|Screwtape removed owner|
|-|-|

Changes made in the attached patch:
 * A couple of instances of "@" have been edited to prevent Sphinx assuming they're email addresses.
 * It turns out that the string `"test_"` (including the double quotes) causes Sphinx to create a hyperlink to nowhere; we now use <code> instead of <q>, which is arguably more correct.
 * "`**kwargs`" is marked as code for consistency with the previous mention.
 * The reference to `__iadd__` is marked as code, because other Python identifiers are so marked.
 * Various lines had trailing or interior whitespace added to avoid confusing Sphinx.
 * The literal markup/example output sections of `doc-standard.xhtml` were given captions to help people distinguish between examples and the actual body of the document; <blockquote> elements were also added for the same reason.
 * A link to "`coding-standard.xhtml#commits`" had the anchor removed because it confused something into making a broken link to "`coding-standard-commits`".
 * A <code> element was removed from inside an <a> element, since nested markup is not supported.

Changes not in the attached patch:
 * There are still frequent unwanted spaces between commas, colons, full-stops and other punctuation.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to Screwtape<br>@exarkun set status to `new`|
|-|-|

  1. doc/core/development/policy/doc-standard.xhtml
    1. Changes look fine.  The lore output is even better than it was.
    1. I noticed a problem with the Sphinx output of "shell" class listings though.  They don't appear to be marked up in any way.  Is this a general problem with lore2sphinx or the stylesheet that we need to correct?
  1. doc/core/development/policy/coding-standard.xhtml looks good
  1. doc/core/development/policy/svn-dev.xhtml
    1. There's still a !``twisted/protocols/imap4.py`` in the output
    1. Missing space: appropriate”test-case-name”
  1. doc/core/development/policy/test-standard.xhtml
    1. I suppose eventually we can have a stdlib module role which is both fixed-width like code and a link.  Does that make sense?  Can you file a ticket for this?
  1. This will also need a news fragment.

**Comment by twisted-trac** [comment]
|<img alt="screwtape's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|screwtape commented|
|-|-|

(In [29790]) Create branch improve-lore2sphinx-output-for-projects-core-development-4567

**Comment by twisted-trac** [comment]
|<img alt="screwtape's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|screwtape commented|
|-|-|

(In [29794]) Applied patch lore-core-development-cleanups.diff from [[#4567](https://github.com/twisted/twisted/issues/4567)](https://github.com/twisted/twisted/issues/4567)

refs #4567

**Comment by twisted-trac** [comment]
|<img alt="screwtape's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|screwtape commented|
|-|-|

(In [29795]) Address exarkun's code-review comments.

refs [#4567](https://github.com/twisted/twisted/issues/4567)

**Comment by twisted-trac** [comment]
|<img alt="Screwtape's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|Screwtape removed owner|
|-|-|

Replying to [exarkun](#note_4):
> I noticed a problem with the Sphinx output of "shell" class listings though.  They don't appear to be marked up in any way.  Is this a general problem with lore2sphinx or the stylesheet that we need to correct?

I dug into this a little, and it seems that lore2sphinx is rendering all three classes ("python", "python-interpreter", "shell") as the ReST equivalent of HTML's <pre> - unadorned, with no class information. However, Sphinx is smart enough to recognise "python" and "python-interpreter" content and syntax-highlight it anyway. Sadly, this cleverness does not extend to shell.

Apparently if lore2sphinx is updated to produce markup like this: 
```
.. code-block:: python
   print "hello world"

.. code-block:: pycon
   >>> 1+1
   2

.. code-block:: console
   $ twistd --version
```

...then the Python examples should look the same, but the shell example should at least be highlighted.

Filed [#4583](https://github.com/twisted/twisted/issues/4583).

> There's still a !``twisted/protocols/imap4.py`` in the output

Fixed.

> Missing space: appropriate”test-case-name”

Fixed.

> I suppose eventually we can have a stdlib module role which is both fixed-width like code and a link.  Does that make sense?  Can you file a ticket for this?

Filed [#4582](https://github.com/twisted/twisted/issues/4582).

>   1. This will also need a news fragment.

**Comment by twisted-trac** [comment]
|<img alt="Screwtape's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|Screwtape commented|
|-|-|

Generated Sphinx docs [here](http://buildbot.twistedmatrix.com/builds/sphinx-html-15623/).

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_10"></a>|@khorn commented|
|-|-|

Replying to [Screwtape](#note_8):
> Replying to [exarkun](#note_4):
> > I noticed a problem with the Sphinx output of "shell" class listings though.  They don't appear to be marked up in any way.  Is this a general problem with lore2sphinx or the stylesheet that we need to correct?
> 
> I dug into this a little, and it seems that lore2sphinx is rendering all three classes ("python", "python-interpreter", "shell") as the ReST equivalent of HTML's <pre> - unadorned, with no class information. However, Sphinx is smart enough to recognise "python" and "python-interpreter" content and syntax-highlight it anyway. Sadly, this cleverness does not extend to shell.
> 
> Apparently if lore2sphinx is updated to produce markup like this: 
> ```
> .. code-block:: python
>    print "hello world"
> 
> .. code-block:: pycon
>    >>> 1+1
>    2
> 
> .. code-block:: console
>    $ twistd --version
> ```
> 
> ...then the Python examples should look the same, but the shell example should at least be highlighted.
> 
> Filed [#4583](https://github.com/twisted/twisted/issues/4583).
> 

FYI, I've made the changes to do this in lore2sphinx.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_12"></a>|@thijstriemstra set owner to Screwtape|
|-|-|

`clients` howto:

 - the last sentence of the 2nd paragraph of the Overview section, it ends with `Protocol .`. That space before the period should go.
 - in the last paragraph of the 'Protocol' section there's a space that needs to go in: ```Protocol -specific ```
 - the `ircLogBot` example has a '../../words/examples/ircLogBot.py' reference, can we get rid of this or turn it into a working link?
 - in the 'Persistent Data in the Factory' section there's a bunch of spaces in the last sentence: '... as self.factory ... case of LogBot , it ...'
 - same issue with the last section of the page, api references have an extra space.

**Comment by twisted-trac** [comment]
|<img alt="Screwtape's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_13"></a>|Screwtape removed owner|
|-|-|

Actually, there's spaces after every instance of inline formatting, whether it be a link or `code` or _emphasis_. It's apparently one of the pitfalls of computer-generated ReST markup that spacing is difficult to get right, and khorn decided to err on the side of 'too much' space rather than 'too little. I mentioned "There are still frequent unwanted spaces" when I put the changes up for review in comment 2; I probably should have repeated that the second time, sorry.

The bare path above included code-listings (such as `ircLogBot.py`) bugs me too, but it seems to be an artifact of the `lore2sphinx` conversion rather than something that we can affect by editing the Lore source. Filed as [#4621](https://github.com/twisted/twisted/issues/4621).

And, uh, now that I think about it, the "clients howto" is part of [#4568](https://github.com/twisted/twisted/issues/4568), not part of the changes in this branch. But thank you for your time, anyway. :)

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_14"></a>|@jml set owner to Screwtape|
|-|-|

In the generated Sphinx documentation:
 * Naming Conventions needs a space in "peer:dataReceived(data)" after the colon.
 * Spurious space in "the documentation standard ," in "Twisted Writing Standard"
 * The "use this template" bit in the "Modules" section of the "Twisted Coding Standard" documentation looks a bit wonky. In particular, the name of the file is just floating above the code listing.

The generated Lore documentation looks fine.

Thanks,
jml

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_15"></a>|@khorn commented|
|-|-|

Replying to [jml](#note_14):

>  * Spurious space in "the documentation standard ," in "Twisted Writing Standard"

This isn't fixable by modifying Lore sources.  It'll need to be fixed up manually.

>  * The "use this template" bit in the "Modules" section of the "Twisted Coding Standard" documentation looks a bit wonky. In particular, the name of the file is just floating above the code listing.
> 

see [#4621](https://github.com/twisted/twisted/issues/4621)

The missing space in the naming Conventinos doc does need to be fixed, though.

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_16"></a>|@khorn set owner to @lvh|
|-|-|

OK, none of the remaining issues can be fixed here, so merge this and we'll fix it manually post-conversion.

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_17"></a>|@khorn commented|
|-|-|

Manual fixes to come under [#4943](https://github.com/twisted/twisted/issues/4943).

**Comment by twisted-trac** [comment]
|[<img alt="lvh's avatar" src="https://avatars.githubusercontent.com/u/97816?s=50" width="50" height="50">](https://github.com/lvh)<a name="note_18"></a>|@lvh set status to `closed`|
|-|-|

(In [31010]) Improved lore2sphinx buildbot results for `projects/core/development`

Author: Screwtape
Reviewer: khorn, exarkun, thijs, jml
Fixes: [#4567](https://github.com/twisted/twisted/issues/4567)

Moving towards Sphinx faster! Yay. Some problems are unfixable, so merging.
